### ISSUE TITLE
linux: handle replay-window incompatible changes between < 6.10 and >= 6.10

### ISSUE THREAD

**Issue by paulwouters** [libreswan/libreswan#1797]
```
               /*
                * Linux kernel >= 6.10 need replay-window 0 on OUTBOUND SA
                * 0. Older kernels does not support replay-window for
                *    OUTBOUND SA with ESN. It support 1.
                * 1. Do BSD varients support 0 on OUTBOUND? If not move next
                *    line to kernel_xfrm.c
                * 2. do we have code to detect Linux kernel version?
                */
               said_next->replay_window = direction == DIRECTION_OUTBOUND ? 0 : c->config->child_sa.replay_window;
```

Should we add a kernel_ops probe for this? Or should we parse/proc/version or use the  uname(2) call for this?

**Comment by cagney** [comment]
Why not leave the code as (but with a comment about how it is nonsensical on outbound):
```
said_next->replay_window = c->config->child_sa.replay_window;
```
and then inside kernel_xfrm.c look at the direction and decide what to do

**Comment by paulwouters** [comment]
sure but that means adding a kernel version check into the code. If we do that, perhaps we should make that available to all parts of pluto ?

This is why I am thinking of uname(2) as that should be more compatible than reading linux specific kernel names/numbers

**Comment by cagney** [comment]
> sure but that means adding a kernel version check into the code.

Such as a call uname(2) in `kernel_xfrm_init()`, sure.  I thought that was the idea.

```
                * 0. Older kernels does not support replay-window for
                *    OUTBOUND SA with ESN. It support 1.
```

Now I'm not sure I follow.

I thought old kernels ignored a non-zero replay-window on outbound while new kernels reject it.
Or is it that old linux kernels required a non-zero replay-window on the outbound state when ESN was enabled?

either way that's a linux quirk.

Hmm, is this so that a config can specify:
```
   esn=yes
   replay-window=0
```
when supported by the kernel?  That would require an addition to the kernel vector.

**Comment by cagney** [comment]
fixed by 9d9b3cb18a48e99a7d01ab1ca0e4473c9ac73a79
